Mathlib/Tactic/NormNum/Basic.lean

```diff
@@ -228,6 +331,68 @@ def evalPow : NormNumExt where eval {u α} e := do
+      have dc : Q(ℕ) := mkRawNatLit qc.den
+      have r1 : Q(Int.pow $na $nb = $nc) := (q(Eq.refl $nc) : Expr)
+      have r2 : Q(Nat.pow $da $nb = $dc) := (q(Eq.refl $dc) : Expr)
+      return (.isRat dα qc nc dc (q(isRat_pow $pa $pb $r1 $r2) : Expr) : Result q($a ^ $b))
   core
+
+theorem isRat_inv_pos {α} [DivisionRing α] : {a : α} → {n d : ℕ} →
+    Nat.ble (nat_lit 1) n = true → IsRat a (.ofNat n) d → IsRat a⁻¹ (.ofNat d) n
+  | _, _, _, _, ⟨_, rfl⟩ => sorry
+
+theorem isRat_inv_zero {α} [DivisionRing α] : {a : α} →
+    IsNat a (nat_lit 0) → IsNat a⁻¹ (nat_lit 0)
+  | _, ⟨rfl⟩ => ⟨by simp⟩
+
+theorem isRat_inv_neg {α} [DivisionRing α] : {a : α} → {n d : ℕ} →
+    Nat.ble 1 n = true → IsRat a (.negOfNat n) d → IsRat a⁻¹ (.negOfNat d) n
+  | _, _, _, _, ⟨_, rfl⟩ => sorry
+
+/-- The `norm_num` extension which identifies expressions of the form `a⁻¹`,
+such that `norm_num` successfully recognises `a`. -/
+@[norm_num _⁻¹] def evalInv : NormNumExt where eval {u α} e := do
+  let .app f (a : Q($α)) ← withReducible (whnf e) | failure
+  let ra ← derive a
+  let dα ← inferDivisionRing α
+  guard <|← withNewMCtxDepth <| isDefEq f q(Inv.inv (α := $α))
+  let ⟨qa, na, da, pa⟩ := ra.toRat'
+  let qb := qa⁻¹
+  if qa > 0 then
+    have lit : Q(ℕ) := na.appArg!
+    let pa : Q(IsRat «$a» (Int.ofNat $lit) $da) := pa
+    let r : Q(Nat.ble (nat_lit 1) $lit = true) := (q(Eq.refl true) : Expr)
+    return (.isRat' dα qb q(.ofNat $da) lit
+      (q(isRat_inv_pos (α := $α) $r $pa) : Expr) : Result q($a⁻¹))
+  else if qa < 0 then
+    have lit : Q(ℕ) := na.appArg!
+    let pa : Q(IsRat «$a» (Int.negOfNat $lit) $da) := pa
+    let r : Q(Nat.ble (nat_lit 1) $lit = true) := (q(Eq.refl true) : Expr)
+    return (.isRat' dα qb q(.negOfNat $da) lit
+      (q(isRat_inv_neg (α := $α) $r $pa) : Expr) : Result q($a⁻¹))
+  else
+    let .isNat inst _z (pa : Q(@IsNat _ AddGroupWithOne.toAddMonoidWithOne $a (nat_lit 0))) := ra
+      | failure
+    return (.isNat inst _z (q(isRat_inv_zero $pa) : Expr) : Result q($a⁻¹))
+
+theorem isRat_div {α} [DivisionRing α] : {a b : α} → {cn : ℤ} → {cd : ℕ} → IsRat (a * b⁻¹) cn cd →
+    IsRat (a / b) cn cd
+  | _, _, _, _, h => by simp [div_eq_mul_inv]; exact h
+
+/-- The `norm_num` extension which identifies expressions of the form `a / b`,
+such that `norm_num` successfully recognises both `a` and `b`. -/
+@[norm_num _ / _, Div.div _ _] def evalDiv : NormNumExt where eval {u α} e := do
+  let .app (.app f (a : Q($α))) (b : Q($α)) ← withReducible (whnf e) | failure
+  let dα ← inferDivisionRing α
+  guard <|← withNewMCtxDepth <| isDefEq f q(HDiv.hDiv (α := $α))
+  let rab ← derive (q($a * $b⁻¹) : Q($α))
+  let ⟨qa, na, da, pa⟩ := rab.toRat'
+  let pa : Q(IsRat ($a * $b⁻¹) $na $da) := pa
+  return (.isRat' dα qa na da q(isRat_div $pa) : Result q($a / $b))
+
+attribute [nolint defLemma] isRat_add isRat_neg isRat_sub isRat_mul isRat_pow isRat_inv_pos
+  isRat_inv_neg isRat_div -- FIXME: figure out how to get the linter to work with these
```

Comment:
```suggestion
```
Maybe `IsRat` should be in `Prop`, just like `IsInt`.  I don't know if we ever need to extract the inverse of the denominator.

But if `IsRat` needs to stay in `Prop`, then these should all be defs and not theorems.  So either way, the nolint needs to be removed.

Reply:
I tried to change as many to `def` as possible, but `isRat_pow` gives an error like "unsupported `Mathlib.Meta.NormNum.IsNat.casesOn` application during code generation" when changed to def, as do the ones in NormNum.Core. What's the fix for this?

Reply:
The fix is to move `IsRat` into `Prop`.  Nothing seems to break because of that and all the theorems can remain theorems.  (Sorries in defs are a big no-no, this will break mathport.)

```lean
inductive IsRat [Ring α] (a : α) (num : ℤ) (denom : ℕ) : Prop
  | mk (inv : Invertible (denom : α)) (eq : a = num * ⅟(denom : α))
```

Reply:
I was waiting for the green check to reply but looks like you beat me to it! Great! :)